sky/jobs/utils.py

```diff
@@ -583,11 +585,22 @@ def should_keep_logging(status: managed_job_state.ManagedJobStatus) -> bool:
                     # Stream the logs to the console without reading the whole
                     # file into memory.
                     start_streaming = False
-                    for line in f:
-                        if log_lib.LOG_FILE_START_STREAMING_AT in line:
-                            start_streaming = True
-                        if start_streaming:
-                            print(line, end='', flush=True)
+                    if tail is not None and tail > 0:
+                        # Read only the last 'tail' lines efficiently using
+                        # deque
+                        lines = collections.deque(f, maxlen=tail)
+                        for line in lines:
+                            if log_lib.LOG_FILE_START_STREAMING_AT in line:
+                                start_streaming = True
+                            if start_streaming:
+                                print(line, end='', flush=True)
+                    else:
+                        # Read all lines
+                        for line in f:
+                            if log_lib.LOG_FILE_START_STREAMING_AT in line:
+                                start_streaming = True
+                            if start_streaming:
+                                print(line, end='', flush=True)
```

Comment:
Alternate suggestion if it works
```suggestion
                    read_from: Union[TextIO, List[str]] = f
                    if tail is not None and tail > 0:
                      # Read only the last 'tail' lines efficiently using
                      # deque
                      read_from = collections.deque(f, maxlen=tail)
                    for line in read_from:
                        if log_lib.LOG_FILE_START_STREAMING_AT in line:
                            start_streaming = True
                        if start_streaming:
                            print(line, end='', flush=True)
```

If it doesn't work feel free to disregard

Reply:
This works and is much more concise. Thank you for the suggestion!